### ISSUE TITLE
Dockerfile support for ENVFILE ./.env?  

### ISSUE THREAD

**Issue by matthewmueller** [docker/docker#28617]
Looked around a bunch but can't find how to do this:

```
ENVFILE ./env
```

If that's not possible, is there a way to build a docker image that will contain all those environment variables when run? 

I see that there's a way to do it when you start a container with docker-run, but I'd like to be able to do it when I build the container.

**Comment by duglin** [comment]
There is no cmd to do that today.  Its an interesting option though.  Do others have a need for this?

**Comment by thaJeztah** [comment]
With the limitations of the env-file in docker (e.g. No multi-line env vars etc), why not use something like;

```Dockerfile
COPY my-env-vars /
RUN export $(cat my-env-vars | xargs) 
```

**Comment by duglin** [comment]
@thaJeztah that will only make those env vars available during the RUN cmd, they won't be persisted in the resulting image, or even in subsequence RUN cmds

**Comment by thaJeztah** [comment]
@duglin ah, you're right

**Comment by ripcurld00d** [comment]
@duglin what about adding this instruction to Dockerfile

```dockerfile
RUN echo "VAR123='myvar_123'" >> /etc/profile.d/myenvvars
```

**NOTE**: Good workaround even thought it doesn't work for all the images.

**Comment by duglin** [comment]
@ripcurld00d yea, I'm sure there are probably work-around based on your image, but I'm not sure why we wouldn't support and env file via build when we already support it via run.  Seems like a nice level of consistency.

**Comment by ssdong** [comment]
@duglin I agree with keeping them consistent. I once had an env variable related issue of trying to import env variables to the container via Dockerfile but couldn't use the built-in `ENV` in Dockerfile because I need dynamic env variables. Fortunately I was using docker-compose and it allows env variables with declaration of `environment` which can read env variable from the host. I would be happy to see this feature added to Dockerfile and keep this consistency with `docker run`. So if Docker support env-file, I might not need to depend on docker compose to pass dynamic env variables. 
@thaJeztah Any thoughts on adding this feature ? 
If it gets approved, can I claim this issue to work on? I've been working with Docker for a few months and love it and would like to make contributions to this great project. 
Thanks!

**Comment by reypm** [comment]
Yeah, I am running exactly into the same issue where `.env` file support will be ideal for this purpose because I need to define almost 20 ENV vars and that will get the Dockerfile very ugly. So any chance for support .env in Dockerfile or BUILD command?

**Comment by tpllaha** [comment]
Ran into the same issue. 
As a workaround I'm doing something like:
```
CMD env `cat <my_env_file> | grep -v '^\s*#'` <my_command>
```
(The grep being there just to ignore comments)

**Comment by mashiox** [comment]
I keep running into this limitation when building images, and would definitely appreciate a `ENFILE` directive. 

I feel like it would make the `docker` subcommands more 1:1 with `docker-compose`

**Comment by TheSeanBrady** [comment]
I'd love this. It will allow people to take an image and extend it with customizable options. A great example of this is an image I use for a binary called "sipp". This binary connects to a PBX and tests it. With an ```ENVFILE``` option I can easily add the PBX's username and password to the image.

**Comment by cpuguy83** [comment]
@TheSeanBrady This wouldn't magically mask the environment variables, the data would still leak into the image.
In such a case you can just use `--build-arg` along with `ARG`.

**Comment by thaJeztah** [comment]
Also sounds like you need the username/password at _runtime_, not during build, in which case `docker run --env-file` would do the trick

**Comment by milky-milk** [comment]
Any news?

**Comment by ilinum** [comment]
I am interested in working on this. Would anyone have any pointers on where to start and how long it would take?

@ssdong did you ever implement this? If you are still interested then you should do it!

**Comment by thaJeztah** [comment]
I'd like to see input from @tonistiigi and @AkihiroSuda first; there's a lot of work being done in BuildKit at this moment (https://github.com/moby/buildkit); also, other than "convenience", interested to hear a bit more about the use-cases for this feature.

**Comment by embray** [comment]
I recently encountered a use case for something like.  There's no great generic way to set a large number of environment variables for all commands run in a container.  For example, while one can set some environment variables in an `/etc/profile` or the like, this is only relevant to running login shells, not arbitrary CMDs.

In my case I script that generates a large number of environment variables that should be set for the application.  I could hard-code those variables with `ENV` commands in the Dockerfile, but it's rather inconvenient because the output of that script could change from version to version of the application. 

It would be nice if (and I'm not sure if this makes sense architecturally) I could do something in a Dockerfile like:

```
RUN myscript.sh > /tmp/environment
ENVFILE /tmp/environment
```

then the build process would actually read the environment from a file *in the image* and, line by line, perform the equivalent `ENV` commands.  Having an `ENVFILE` command that can read a file in the image is nice because it also covers cases like

```
COPY .envs /tmp/.envs
ENVFILE /tmp/.envs
```

in that case it becomes a two step process, but it also affords quite a bit of flexibility.

**Comment by embray** [comment]
Also, I realize I could achieve the equivalent effect using an `ENTRYPOINT`, but at least in my case that slows down any command run with that image, since the script that generates the environment is non-trivial.

**Comment by thiagolsfortunato** [comment]
Any news?

**Comment by vsoch** [comment]
I'd like to see support for this as well, and can add a use case! I'm writing a script to convert a Singularity image to Docker, and Singularity has a basic text file with a bunch of environment settings in it. I either need to parse that manually (annoying but possible!) or I could just specify a path to the file during build.

TLDR :+1: !

**Comment by lillypad** [comment]
Bump!

**Comment by ticklemynausea** [comment]
☹️

**Comment by rm-rf-etc** [comment]
I don't know why we need more use cases to support this. It's common to have files for different environments, e.g., env_dev, env_staging, env_prod. I'm still new-ish to docker/docker-compose + kubernetes, and I'm confused as to what the official workflow is supposed to be. Am I supposed to create a Dockerfile for each environment? There's `docker-compose build` but when run, it complains that I don't have a Dockerfile, so here I am...

**Comment by isaiahgrant** [comment]
Another use case: Building an application with the application configuration done at build-time.

In my case the build configuration creates files accessible by the build but not by the application (aws-cli for example). It also passes through configurations to the .env file that is used by the application itself.

If these configurations were made available at run-time (using `docker run --env-file ...`) the CMD would be something along the lines of:
`CMD ["python", "app/handle_env_vars_and_configure_application.py"]`
Where the CMD here would then chain (run) the command that starts the application once it is done building configuration.

Alternatively with a build env file it is simply:
`CMD ["python", "app/run_my_application.py"]`
Since all configuration has happened before this point. Also, when running the container later only run-time variables need to be passed separating those from other environment (build-time) variables.

Finally:

`docker build --build-arg CONFIG_1=1 --build-arg CONFIG_2=2 --build-arg CONFIG_3=3 .... --build-arg CONFIG_X=X`

Gets quite unwieldy and tedious to manage after only a handful of variables. Production vs. Staging build commands are complex and potentially largely different (rather than simply providing different ENV files).

I feel like I could keep going on this point, but I hope this is enough in addition the other valid concerns placed (well) before mine :).

:+1:

**Comment by huggla** [comment]
You can preserve environment variables from previous build stages with this little trick:
```
FROM whatever as stage1
ARG ENTRYPOINT="/usr/bin/entry"
ENV STAGE1VAR="test"
RUN { echo '#!/bin/sh' && export -p && echo "exec $ENTRYPOINT "'"$@"'; } > /usr/bin/env-entrypoint \
 && chmod +x /usr/bin/env-entrypoint
  
FROM somethingelse
COPY --from=stage1 /usr/bin/env-entrypoint /usr/bin/
ENTRYPOINT ["/usr/bin/env-entrypoint"]
CMD ["/usr/bin/mycommand", "arg"]
```

**Comment by colegatron** [comment]
Be able to specify env files at build time is a must, IMHO.
We would archieve multiple goals:
- Set large amounts of environment variables
- Use that env vars for LABELS. 
- Remove the dependency files to be stored in a specific build server.
  Taking advantage docker's feature of sending the build context to the build server ( -H thisserver ) 
 you can bundle any required file in the build through ENV vars: 
`$ MYCFG=$( base64 -i stagging.json )`
`$ DEPFILE1=$( base64 -i tomcat.xml )`
`$ DEPFILE2=$( base64 -i whatever.yaml )`
`$ docker build --build-arg CFG=$MYCFG --build-arg $DEPFILE1  apibuilder`

Preparing an .env-file with whatever files/config you need then you can ask docker to run the build in any server you want, without being tied to you 'build server'.
Also, avoiding to must have a powerful build server reusing spare cpu cycles on other idle servers (or even workstations) can save you a lot of $$$ in cloud bills.

**Comment by huggla** [comment]
You could format your envfile like this:
```
#!/bin/sh

ENTRYPOINT="/usr/bin/your_entrypoint"

export \
   VAR1=value1 \
   VAR2=value2 \
   VAR3=value3

exec $ENTRYPOINT "$@"
```
Then near the end of your dockerfile add:
```
COPY ./envfile /envfile
RUN chmod +x /envfile
ENTRYPOINT ["/envfile"]
```

**Comment by huggla** [comment]
You are absolutely right, except for the process reaping part. The exec last in my script makes whatever command that is in $ENTRYPOINT PID 1.

**Comment by colegatron** [comment]
True, exec in shell replaces process without forking.
Confused the exec in the script with dockerfile commands

**Comment by r3c4ll** [comment]
No solution for this yet?

**Comment by valentinradu** [comment]
This would be great!

**Comment by GJRTimmer** [comment]
I really would to be able to specify a env_file during build.

**Comment by FisherKK** [comment]
Hello I would also welcome this feature.

**Comment by asoltesz** [comment]
We very much need this because there is a pretty big disparity now with Docker Compose regarding this. 

Also, the possible workarounds are not very nice.

**Comment by espetro** [comment]
Hey there,
It would pretty dope to have this for Docker builds! Looking forward to it 😄

**Comment by hugoalvarado** [comment]
I had a need for something similar for making env variables available to a cron job, and I ended up using this:
```ADD .env /code
RUN cat /code/.env >> /etc/environment```

**Comment by rulatir** [comment]
All the answers that involve workarounds are misguided. We shouldn't **need** workarounds. What we need is full expressive power of key-value pairs, no limitations, no reservations, always, everywhere, no exceptions, no compromises. Passing values from point A to point B in a Dockerfile should be as natural as breating rather than a source of constant asphyxiation and a PROBLEM that generates VOLUMES of folk medicine on how to cure it in every last one of the gazilliard corner cases.

**Comment by elfro** [comment]
Any updates?

**Comment by zulrang** [comment]
How has this been open and untouched for 2.5 years?

**Comment by cpuguy83** [comment]
@zulrang Apparently no one has thought it important enough to take the time to design and implement it.

**Comment by vsoch** [comment]
As a workaround you can use a `docker-compose.yml` file to wrap running the container, something like:
```bash
version: '2'
services:
  postgres:
    image: postgres:10
    environment:
      - POSTGRES_USER=postgres
    env_file:
      - .env
```
In the above, we define both environment variables that aren't secrets under `environment` (the docker-compose might be shared on GitHub, for example) along with secrets in `.env`. Then you'd start:

```bash
$ docker-compose up -d
```
The issue has been open for years because it's obviously not important enough to warrant anyone looking at. What choice do we have then? We can keep posting here, implement a pull request to implement the suggested change (and hope someone will look at it) or just come up with another method to achieve similar. Not great options, but what else can you do?

**Comment by AkihiroSuda** [comment]
It could be experimented as a custom Dockerfile frontend for BuildKit

**Comment by SantoshSah** [comment]
any update?

**Comment by alexkreidler** [comment]
Now that multi-stage builds are a prevalent pattern, it is really important to have configuration options when running the first stages. 

If someone were to create a PR for this, would the Docker/Moby team be interested in adding this functionality?

**Comment by gaui** [comment]
Why isn't this higher in priority?

**Comment by sumitcn** [comment]
any update ?

**Comment by zhule0601** [comment]
any new ?

**Comment by rduzagac-ebay** [comment]
any update?

**Comment by helzgate** [comment]
For now, I got rid of my stages and went to a single bloated image.

**Comment by vsoch** [comment]
If anyone is interested in a read only container (or uses HPC) both `--env` and `--env-file` are being added to Singularity - you can review the [PR here](https://github.com/sylabs/singularity/pull/5028). So - if you have a Dockerfile that builds a scientific container and then you put it on Docker Hub, you could pull into a Singularity container and then use with the external file.

**Comment by pietro-tavares** [comment]
Is a PR for this welcome?

**Comment by AkihiroSuda** [comment]
Should be experimented as a custom LLB frontend first.

**Comment by pietro-tavares** [comment]
@AkihiroSuda Sure. Is there any specific guideline I should follow?

Seems to me that [the BuildKit documentation](https://github.com/moby/buildkit#exploring-llb) suffices.

**Comment by AkihiroSuda** [comment]
* Add a Dockerfile instruction parser to https://github.com/moby/buildkit/tree/master/frontend/dockerfile/instructions with custom build tag  like `+ build dfenvfile`
* Implement an LLB generator as well in https://github.com/moby/buildkit/tree/master/frontend/dockerfile/dockerfile2llb
* Build the frontend image using https://github.com/moby/buildkit/blob/master/frontend/dockerfile/cmd/dockerfile-frontend/Dockerfile
* If it works, feel free to propose the work to the upstream in https://github.com/moby/buildkit/issues

**Comment by AkihiroSuda** [comment]
@tonistiigi : Does this proposal SGTY?

**Comment by CloudZazu** [comment]
I ran into the need for this

**Comment by stanislav-yaranov** [comment]
still nothing after almost 4 years?

**Comment by thaJeztah** [comment]
_If_ this would be implemented, we need to look carefully at the design. The `--env-file` option on `docker run` has many limitations (which should be looked at if we want to inherit the "bad parts"), e.g.

- current implementation has no support for multi-line values
- current implementation does not support `export` lines. While this was by design, some users indicated that it limits re-use of the same file for multiple purposes
- should env-vars be expanded based on the environment on the host (like on `docker run`), or expanded based on the "build" container/environment?

For example;

```bash
cat envfile
SOME_VAR
HELLO=world
FOO=$HELLO
EMPTY_VALUE=
NO_SUCH_ENV_IN_ENVIRONMENT

docker run --rm --env-file=./envfile busybox env
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=aa0482d94952
SOME_VAR=I am some var
HELLO=world
FOO=$HELLO
EMPTY_VALUE=
HOME=/root
```

Having said the above, I think implementing `--env-file` would only fixes _one_ option that can be tedious to type. There are now many options available for `docker build` (build args, network, dns-settings, labels, quota, security options, ...), and I think that a `docker-compose.yml` or [`docker buildx bake` with a bake-file](https://github.com/docker/buildx#buildx-bake-options-target) to be better options to look into.

There's still improvements that can be made in that area, such as native support for `docker-compose` on `docker build` (currently already supported by `docker buildx`), and the compose-specification may need improvements to make `build` more of a first-class citizen (something to be discussed in https://github.com/compose-spec/compose-spec)

**Comment by jayamorin** [comment]
I'm still waiting and can't figure a workaround.

**Comment by Elliot9** [comment]
i use docker-compose load .env file and pass the args to dockerfile, then setting arg ,env  in dockerfile.
 it's a stupid way, but solve my problems.

For example:

`
#in yml
services:
  nginx:
    build:
      context: .
      dockerfile: dockerfile
      args:
        Project: ${Project}
`

`
#dockerfile
FROM nginx:latest
ARG Project
ENV Project=${Project}
...
...
CMD envsubst '${Project}' < /etc/nginx/conf.d/default.template > /etc/nginx/conf.d/default.conf && exec nginx -g 'daemon off;'
`

**Comment by pablodz** [comment]
Why is not implemented yet?

**Comment by 2kCarlos** [comment]
Still waiting in 2022..

This would greatly help me as well.

**Comment by colegatron** [comment]
4 years after commenting to this issue, it is still open, like if it was not important for Docker. 
Now that everybody is moving in k8s to containerd, maybe it is time to do not expect anything else from Docker?

I am using kubernetes since back 2016, but there are still scenarios where this would be nice to have this solution.
Maybe we need to move into Kaniko + Nomad for this kind of scenarios and finally abandon Docker?

**Comment by uniquestring** [comment]
> Now that everybody is moving in k8s to containerd, maybe it is time to do not expect anything else from Docker?
>
> I am using kubernetes since back 2016, but there are still scenarios where this would be nice to have this solution.
> Maybe we need to move into Kaniko + Nomad for this kind of scenarios and finally abandon Docker?

Just like other software solutions, "even" Docker supports env files at runtime. So I don't think you need to compare with k8s or nomad. As I understand this issue, it is only about supporting env files at image build time (i.e. extending Dockerfile syntax).

I have no experience with Kaniko, but from what I see, it looks like they are just trying to be compatible with Dockerfiles. So I'm not sure if they would add this functionality on their end. They would need to "fork" the official Dockerfile syntax, extend it, and by that break compatibility between the two, just for some custom knickknacks. Sound like a bigger hassle than it's worth.

Considering the idle time, and that this feature might never be implemented, in my opinion, your best bet is to just generally look for a building tool that is able to create OCI compliant images. Then choose the one that best fits your needs and roll with it.

**Comment by colegatron** [comment]
So we should not expect this to be implemented.

Thanks.

**Comment by uniquestring** [comment]
> So we should not expect this to be implemented.
>
> Thanks.

I can be mistaken, but to me, your answer sounds like you could have taken my answer as an official statement, that this feature will not be implemented.

Just to be clear: I'm not involved into the Docker development at all. I'm just a random person that wrote their personal view on this.

**Comment by colegatron** [comment]
Thanks for the clarification.
